x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java

```diff
@@ -374,20 +374,29 @@ private void startComputeOnRemoteClusters(
         var linkExchangeListeners = ActionListener.releaseAfter(computeListener.acquireAvoid(), exchangeSource.addEmptySink());
         try (RefCountingListener refs = new RefCountingListener(linkExchangeListeners)) {
             for (RemoteCluster cluster : clusters) {
+                final String clusterAlias = cluster.clusterAlias();
+                final boolean suppressRemoteFailure = computeListener.shouldIgnoreRemoteErrors(clusterAlias);
+                final var exchangeListener = refs.acquire();
                 ExchangeService.openExchange(
                     transportService,
                     cluster.connection,
                     sessionId,
                     queryPragmas.exchangeBufferSize(),
                     esqlExecutor,
-                    refs.acquire().delegateFailureAndWrap((l, unused) -> {
-                        var remoteSink = exchangeService.newRemoteSink(rootTask, sessionId, transportService, cluster.connection);
+                    ActionListener.wrap(unused -> {
+                        var remoteSink = exchangeService.newRemoteSink(
+                            rootTask,
+                            sessionId,
+                            transportService,
+                            cluster.connection,
+                            suppressRemoteFailure ? ex -> computeListener.markAsPartial(clusterAlias, ex) : null
```

Comment:
I'm confused - why does this failureCollector need to mark exceptions as partial but also in the exception handler to this same ActionListener.wrap you also have a block that catches exceptions and marks as partial. When does one fire vs. the other?

We'll probably need to at a minimum document that better and preferably build a code flow or data structure that makes it clear what is happening.

Reply:
There are different failures - one handles the failure in the sink (that's the one on line 392), and another one below handles the failure for the exchange. So these relate to separate listeners which pick up failures in separate parts of the chain resulting from different tasks/requests. We could make a walkthrough on Tuesday if you want, this code is really hairy...